src/components/flame-graph/Canvas.js

```diff
@@ -48,6 +49,138 @@ const ROW_HEIGHT = 16;
+}
+
+/**
+ * Return the background color used to paint a flame graph box. The
+ * passed stack type determines the hue while the self time determines
+ * the lightness of the color. Darker colors indicate higher self
+ * times.
+ */
+export function getBackgroundColor(
+  stackType: StackType,
+  selfTimeRelative: number
+): string {
+  /**
+   * The colors are defined as a range where the starting color is the
+   * color shown when the function has no self time at all, and the
+   * ending color is when all available self time (the total time) in
+   * the flame graph is assigned to a single function. This latter
+   * case occurs when there's only a single box at the top, stretching
+   * over all horizontal space in the graph, covering all boxes below
+   * it. When the self time lies in between these two extremes, the
+   * color is an interpolated value lying between the starting and
+   * ending colors in the range.
+   *
+   * The interpolated color is derived from `selfTimeRelative`, which
+   * is the self time of the frame relative to all available self time
+   * possibly represented in the flame graph. However, using this
+   * value to interpolate to a color linearly will require the value
+   * to be relatively big in order to be able to see the shift in
+   * color, and such large self times would be visually apparent from
+   * the shape of the flame graph anyway. We are better off if we can
+   * discern also small values of self time from the color, and this
+   * is accomplished by changing the linear interpolation to a
+   * logarithmic one instead, with a steep increase in the beginning.
+   *
+   * Below we use the transformation y(x) = log(c * x + 1) / log(c + 1),
+   * where c is a constant. It has the property that y(0) = 0 and
+   * y(1) = 1. The constant has been chosen by visual inspection of
+   * the graph.
+   */
+
+  let backgroundRangeStart;
+  let backgroundRangeEnd;
+
+  const t = Math.log(5000 * selfTimeRelative + 1) / Math.log(5001);
+
+  switch (stackType) {
+    case 'native':
+      backgroundRangeStart = NATIVE_START_COLOR;
+      backgroundRangeEnd = NATIVE_END_COLOR;
+      break;
+    case 'js':
+      backgroundRangeStart = JS_START_COLOR;
+      backgroundRangeEnd = JS_END_COLOR;
+      break;
+    case 'unsymbolicated':
+      backgroundRangeStart = UNSYMBOLICATED_START_COLOR;
+      backgroundRangeEnd = UNSYMBOLICATED_END_COLOR;
+      break;
+    default:
+      throw new Error(`Unknown stack type case "${(stackType: empty)}".`);
+  }
+  return lerpColors(backgroundRangeStart, backgroundRangeEnd, t);
+}
+
+/**
+ * Return the foreground color used to draw text in a flame graph
+ * box. The color matches the background color returned from
+ * `getBackgroundColor` and this function thus expects the same
+ * parameters.
+ */
+export function getForegroundColor(
+  stackType: StackType,
+  selfTimeRelative: number
+): string {
+  const t = Math.log(5000 * selfTimeRelative + 1) / Math.log(5001);
```

Comment:
How expensive are function calls? I skipped extracting this formula into its own function.

Reply:
In unoptimized code, there is a small cost to call out to other functions, but optimized functions will often inline functions and remove that cost. The risk on using lots of functions, is that if the call uses differently shaped arguments, it can be de-optimized. This starts to get into the realm of micro-optimizations though. The key is to profile and see what is going on. The stack chart will show if the code was optimized or not.

Reply:
Alright, thanks for the answer.